src/app/util/ember-compatibility-functions.cpp

```diff
@@ -285,39 +285,62 @@ class MandatoryGlobalAttributeReader : public AttributeAccessInterface
 };
 
-class AttributeListReader : public MandatoryGlobalAttributeReader
+class GlobalListAttributeReader : public MandatoryGlobalAttributeReader
 {
 public:
-    AttributeListReader(const EmberAfCluster * aCluster) : MandatoryGlobalAttributeReader(aCluster) {}
+    GlobalListAttributeReader(const EmberAfCluster * aCluster) : MandatoryGlobalAttributeReader(aCluster) {}
 
     CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
 };
 
-CHIP_ERROR AttributeListReader::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+CHIP_ERROR GlobalListAttributeReader::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
 {
-    // The id of AttributeList is not in the attribute metadata.
+    using namespace Clusters::Globals::Attributes;
+    // The id of the attributes below is not in the attribute metadata.
     // TODO: This does not play nicely with wildcard reads.  Need to fix ZAP to
     // put it in the metadata, or fix wildcard expansion to add it.
-    return aEncoder.EncodeList([this](const auto & encoder) {
-        constexpr AttributeId ourId = Clusters::Globals::Attributes::AttributeList::Id;
-        const size_t count          = mCluster->attributeCount;
-        bool addedOurId             = false;
-        for (size_t i = 0; i < count; ++i)
-        {
-            AttributeId id = mCluster->attributes[i].attributeId;
-            if (!addedOurId && id > ourId)
+    switch (aPath.mAttributeId)
+    {
+    case AttributeList::Id:
+        return aEncoder.EncodeList([this](const auto & encoder) {
+            constexpr AttributeId ourId = Clusters::Globals::Attributes::AttributeList::Id;
+            const size_t count          = mCluster->attributeCount;
+            bool addedOurId             = false;
+            for (size_t i = 0; i < count; ++i)
+            {
+                AttributeId id = mCluster->attributes[i].attributeId;
+                if (!addedOurId && id > ourId)
+                {
+                    ReturnErrorOnFailure(encoder.Encode(ourId));
+                    addedOurId = true;
+                }
+                ReturnErrorOnFailure(encoder.Encode(id));
+            }
+            if (!addedOurId)
             {
                 ReturnErrorOnFailure(encoder.Encode(ourId));
-                addedOurId = true;
             }
-            ReturnErrorOnFailure(encoder.Encode(id));
-        }
-        if (!addedOurId)
-        {
-            ReturnErrorOnFailure(encoder.Encode(ourId));
-        }
+            return CHIP_NO_ERROR;
+        });
+    case ClientGeneratedCommandList::Id:
+        return aEncoder.EncodeList([this](const auto & encoder) {
+            for (size_t i = 0; mCluster->clientGeneratedCommandList && mCluster->clientGeneratedCommandList[i] != 0xFFFF'FFFF; ++i)
+            {
+                ReturnErrorOnFailure(encoder.Encode(mCluster->clientGeneratedCommandList[i]));
```

Comment:
How about:
```suggestion
            for (CommandId * cmd = mCluster->clientGeneratedCommandList; cmd && *cmd != kInvalidCommandId; ++cmd)
            {
                ReturnErrorOnFailure(encoder.Encode(*cmd);
```
and add kInvalidCommandId to DataModelTypes.h.

Reply:
Updated